File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        if(nums.length==0){
            return new int[]{-1,-1};
        }
        int first=firstOccurence(nums,target);
        int last=lastOccurence(nums,target);
        return new int[]{last,first};
    }
    public int firstOccurence(int[] arr,int target){
        int low=0;
        int high=arr.length-1;
        int result=-1;
        while(low<=high){
            int mid=low+(high-low)/2;
            if(arr[mid]==target){
                result=mid;
                low=mid+1;//right
            }
            else if(arr[mid]>target){
                high=mid-1;
            }
            else{
                low=mid+1;
            }
        }
        return result;

    }
    public int lastOccurence(int[] arr,int target){
        int low=0;
        int high=arr.length-1;
        int result=-1;
        while(low<=high){
            int mid=low+(high-low)/2;
            if(arr[mid]==target){
                result=mid;
                high=mid-1;//right
            }
            else if(arr[mid]>target){
                high=mid-1;
            }
            else{
                low=mid+1;
            }
        }
        return result;

    }
}

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    int first(vector<int>& arr,int target){
        int n=arr.size();
        int low=0;
        int high=n-1;
        int mid=-1;
        int ans=-1;
        while(low<=high){
            mid=low+(high-low)/2;
            if(arr[mid]==target){
                ans=mid;
                high=mid-1;
            }
            else if(arr[mid]>target){
                high=mid-1;
            }
            else{
                low=mid+1;
            }
        }
        return ans;

    }
    int last(vector<int>& arr,int target){
        int n=arr.size();
        int low=0;
        int high=n-1;
        int mid=-1;
        int ans=-1;
        while(low<=high){
            mid=low+(high-low)/2;
            if(arr[mid]==target){
                ans=mid;
                low=mid+1;
            }
            else if(arr[mid]>target){
                high=mid-1;
            }
            else{
                low=mid+1;
            }
        }
        return ans;

    }
    vector<int> searchRange(vector<int>& nums, int target) {
        // if(nums.size()==1){
        //     return {0,0};

        // }
        vector<int>arr;
        arr.push_back(first(nums,target));
        arr.push_back(last(nums,target));
        return arr;
        
    }
};